OpenRA.Game/Network/LocalizedMessage.cs

```diff
@@ -0,0 +1,119 @@
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.RegularExpressions;
+using Fluent.Net;
+using OpenRA.Widgets;
+
+namespace OpenRA.Network
+{
+	public class FluentArgument
+	{
+		[Flags]
+		public enum FluentArgumentType
+		{
+			String = 0,
+			Number = 1,
+		}
+
+		public readonly string Key;
+		public readonly string Value;
+		public readonly FluentArgumentType Type;
+
+		public FluentArgument() { }
+
+		public FluentArgument(string key, object value)
+		{
+			Key = key;
+			Value = value.ToString();
+			Type = value is sbyte || value is short || value is int || value is long ||
+				value is byte || value is ushort || value is uint || value is ulong ||
+				value is float || value is double || value is decimal ?
+				FluentArgumentType.Number : FluentArgumentType.String;
+		}
+	}
+
+	public class LocalizedMessage
+	{
+		public const int ProtocolVersion = 1;
+
+		public readonly string Key;
+
+		[FieldLoader.LoadUsing(nameof(LoadArguments))]
+		public readonly FluentArgument[] Arguments;
+
+		static object LoadArguments(MiniYaml yaml)
+		{
+			var arguments = new List<FluentArgument>();
+			var argumentsNode = yaml.Nodes.FirstOrDefault(n => n.Key == "Arguments");
+			if (argumentsNode != null)
+			{
+				var regex = new Regex(@"Argument@\d+");
+				foreach (var argument in argumentsNode.Value.Nodes)
+					if (regex.IsMatch(argument.Key))
+						arguments.Add(FieldLoader.Load<FluentArgument>(argument.Value));
+			}
+
+			return arguments.ToArray();
+		}
+
+		static readonly string[] SerializeFields = { "Key" };
+
+		public LocalizedMessage(MiniYaml yaml)
+		{
+			FieldLoader.Load(this, yaml);
+		}
+
+		public LocalizedMessage(string key, Dictionary<string, object> arguments = null)
+		{
+			Key = key;
+			Arguments = arguments?.Select(a => new FluentArgument(a.Key, a.Value)).ToArray();
+		}
+
+		public string Serialize()
+		{
+			var root = new List<MiniYamlNode>() { new MiniYamlNode("Protocol", ProtocolVersion.ToString()) };
+			foreach (var field in SerializeFields)
+				root.Add(FieldSaver.SaveField(this, field));
+
+			if (Arguments != null)
+			{
+				var argumentsNode = new MiniYaml("");
+				var i = 0;
+				foreach (var argument in Arguments)
+					argumentsNode.Nodes.Add(new MiniYamlNode("Argument@" + i++, FieldSaver.Save(argument)));
+
+				root.Add(new MiniYamlNode("Arguments", argumentsNode));
+			}
+
+			return new MiniYaml("", root)
+				.ToLines("LocalizedMessage")
+				.JoinWith("\n");
+		}
+
+		public string Translate()
+		{
+			var argumentDictionary = new Dictionary<string, object>();
+			foreach (var argument in Arguments)
+			{
+				if (argument.Type == FluentArgument.FluentArgumentType.Number)
+					argumentDictionary.Add(argument.Key, new FluentNumber(argument.Value));
+				else
+					argumentDictionary.Add(argument.Key, new FluentString(argument.Value));
```

Comment:
`var type = argument.Type == FluentArgument.FluentArgumentType.Number ? new FluentNumber(argument.Value) : new FluentString(argument.Value);`

Reply:
> Type of conditional expression cannot be determined because there is no implicit conversion between 'Fluent.Net.FluentNumber' and 'Fluent.Net.FluentString' [OpenRA.Game]csharp(CS0173)

Reply:
Hah, ofc!